specifyweb/frontend/js_src/lib/components/QueryBuilder/CheckReadAccess.tsx

```diff
@@ -0,0 +1,66 @@
+import React from 'react';
+import { commonText } from '../../localization/common';
+import { queryText } from '../../localization/query';
+import { RA } from '../../utils/types';
+import { Ul } from '../Atoms';
+import { SerializedResource } from '../DataModel/helperTypes';
+import { SpQuery, Tables } from '../DataModel/types';
+import { Dialog } from '../Molecules/Dialog';
+import { TableIcon } from '../Molecules/TableIcon';
+import { hasTablePermission } from '../Permissions/helpers';
+import { QueryFieldSpec } from './fieldSpec';
+
+export function CheckReadAccess({
+  query,
+}: {
+  readonly query: SerializedResource<SpQuery>;
+}): JSX.Element | null {
+  const [noAccessTables, setNoAccessTables] = React.useState<RA<keyof Tables>>(
+    []
+  );
+
+  function accessTables(query: SerializedResource<SpQuery>): RA<keyof Tables> {
+    const info = query.fields.flatMap((field) => {
+      const fieldSpec = QueryFieldSpec.fromStringId(
+        field.stringId,
+        field.isRelFld ?? false
+      );
+      const relatedTable = fieldSpec.joinPath.map(({ model }) => model.name);
+      return relatedTable;
+    });
+
+    const withoutDuplicates = new Set(info);
+
+    const noAccessTables = Array.from(withoutDuplicates).filter(
+      (name) => !hasTablePermission(name, 'read')
+    );
+
+    setNoAccessTables(noAccessTables);
+    return info;
+  }
+  React.useEffect(() => {
```

Comment:
Try to avoid useEffect whenever possible. In this case, useMemo would work better. Or even simpler
```js
const [noAccessTables,setNoAccessTables] = React.useState(getNoAccessTables);
```

Where `getNoAccessTables` is a function that would return `RA<keyof Tables>`

Reply:
This works because you can provide the initial value to React.useState as the value itself, **or as a function that returns default value**